packages/langium/src/generator/generator-node.ts

```diff
@@ -78,6 +97,31 @@ export class CompositeGeneratorNode {
         return this.contents.length === 0;
     }
 
+    /**
+     * Adds tracing information to `this` generator node. Overwrites existing trace data, if set previously.
+     *
+     * The given data are kept as they are, the actual resolution of text positions within the DSL text
+     * is done at the final processing of `this` node as part of {@link toStringAndTrace()}.
+     *
+     * @param astNode the AstNode corresponding to `this` node's content
+     *
+     * @param property the value property name (string) corresponding to `this` node's content,
+     *  e.g. if this node's content corresponds to some `string` or `number` property; is optional
+     *
+     * @param index the index of the value within a list property corresponding to `this` node's content,
+     * if the property contains a list of elements, is ignored otherwise,
+     * must not be given if no `property` is given ; is optinal
+     *
+     * @returns `this` {@link CompositeGeneratorNode} for convenience.
+     */
+    trace<T extends AstNode>(astNode: T, property?: Properties<T>, index?: number): this {
```

Comment:
It would be cool to be able to trace to a CST node as alternative (use two different method signatures).

Same for the other trace methods.

Reply:
What do you have in mind? Being able to provide a basically arbitrary text region?

Reply:
> Trace things that are not covered by a property. Examples: comments, unassigned keywords.

Indeed. I'm just thinking whether we can add this later without breaking anything.

Reply:
You'd have to change the `TraceSourceSpec` type, which might be breaking.

I'm mentioning this because we have seen quite extensive use of the CST (Node Model) with Xtext in earlier projects, including large-scale generators. Of course we can argue that if you need the information in a generator, you should assign it to an AST property. So yes, we can wait for a concrete case where we need it, even if it partly breaks API.

Reply:
>I'm mentioning this because we have seen quite extensive use of the CST (Node Model) with Xtext in earlier projects

It's fine, I've added that as you're totally right; implementation is almost done ... will push another update soon ;-)

Reply:
I agree with @spoenemann here. Basically all APIs that we have that are somehow related to pointing to a certain range usually accept properties, cst nodes and ranges. [The semantic token provider](https://github.com/langium/langium/blob/c7e15ee9663d37dab676dea4ffce479e3af0fea3/packages/langium/src/lsp/semantic-token-provider.ts#L79-L98) shows this quite well.

Reply:
This test case illustrates one example of what I can imagine to be relevant in practice: https://github.com/langium/langium/blob/bb5d1800fc850013ba5a023035d2564f0ec5d851/packages/langium/test/generator/generation-tracing.test.ts#L147-L170